### PR TITLE
feat(collection): New basic collection types.

### PR DESCRIPTION
**Pull request by ttowncompiled**
Adds four basic collection types for better type enforcement on collections.
Assortment: <key: number || string>, <value: any>
Collection: <key: number || string>, <value: T>
Dictionary: <key: string>, <value: any>
Thesaurus: <key: string>, <value: T>

### CODE REVIEW

**Comment by ttowncompiled** [comment]
I did want to note that I created these types because I have been in multiple situations thus far where I have defined types that were very similar to these. I thought that others might be running into the same redundancies and that they might find it nice to have a basic set of types in the DefinitelyTyped repository that they can lean on.

Also, for syntax in the commit: string || number means could be string or could be number, either works. Not to be confused with sub-union types, though it sub-union types do get created, they would be equivalent to what I'm trying to convey.

**Comment by vvakame** [comment]
*collection/collection.d.ts*

check list

* [ ] is collect [naming convention](http://definitelytyped.org/guides/contributing.html#naming-the-file)?
  * https://www.npmjs.com/package/collection
  * http://bower.io/search/?q=collection
  * others?
* [X] has a [test file](http://definitelytyped.org/guides/contributing.html#tests)? (collection/collection-tests.ts or others)
* [ ] pass the Travic-CI test?

File: collection/collection.d.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Type definitions for collections
+// Project: https://github.com/ttowncompiled/DefinitelyTyped
```
Comment (vvakame) [inline collection/collection.d.ts]: is this right? https://github.com/tfeng/collection ?

**Comment by ttowncompiled** [comment]
Not quite. Though, I will probably type that one up later.

Thanks for mentioning that though because I should probably rename this now.

I was hoping to provide a very basic set of collection types that others could reference and use. For instance, I just finished typing up protobufjs, and there was a number of instances where the type was a StringMap that could contain any value so having the Dictionary type provided a much more convenient means of describing those objects.

I though it would be really nice to have a directory inside DefinitelyTyped that would provide a multitude of different common data structures.

Is that something that would be of interest?

**Comment by ttowncompiled** [comment]
Okay. I changed the name of the directory so that it won't conflict with any npm packages. Just to reiterate, right now these collections are here mostly for conveyance. Seeing a Dictionary type for instance will tell inform whoever is using the object that all keys will be strings, even though there is no TS enforcement to ensure that.

Hopefully, in the future there will be though and then these provided types will actually be very useful.

**Pull request closed without merging**